src/hotspot/share/gc/shared/preservedMarks.cpp

```diff
@@ -34,21 +34,24 @@
 
 void PreservedMarks::restore() {
   while (!_stack.is_empty()) {
-    const OopAndMarkWord elem = _stack.pop();
+    const PreservedMark elem = _stack.pop();
     elem.set_mark();
   }
   assert_empty();
 }
 
+void PreservedMarks::adjust_preserved_mark(PreservedMark* elem) {
+  oop obj = elem->get_oop();
+  if (obj->is_forwarded()) {
```

Comment:
Pre-existing: we probably only need to check this because G1 full gc preserves marks during marking, unlike the other collectors which (attempt to) preserve marks for actually forwarded objects.
I'll file an RFE for this.